EIPS/eip-4788.md

```diff
@@ -52,43 +55,75 @@ Validity is guaranteed from the consensus layer, much like how withdrawals are h
 At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions),
 write the parent beacon root provided in the block header into the storage of the contract at `HISTORY_STORAGE_ADDRESS`.
 
-The root itself is used as a key into the contract's storage and the timestamp of the header is written as the key's value.
-The timestamp (a 64-bit unsigned integer value) is encoded as 32 bytes in big-endian format.
+In order to bound the storage used by this precompile, two ring buffers are used: one to track the latest timestamp at a given index in the ring buffer and another to track
+the latest root at a given index.
+
+To derive the index `timestamp_index` into the timestamp ring buffer, the timestamp (a 64-bit unsigned integer value) is reduced modulo `HISTORICAL_ROOTS_LENGTH`.
+To derive the index `root_index` into the root ring buffer, add `HISTORICAL_ROOTS_LENGTH` to the index into the timestamp ring buffer.
+Both resulting 64-bit unsigned integers should be encoded as 32 bytes in big-endian format when writing to the storage.
+
+The timestamp from the header, encoded as 32 bytes in big-endian format, is the value to write behind the `timestamp_index`.
+The 32 bytes of the `parent_beacon_block_root` (as provided) are the value to write behind the `root_index`.
 
 In Python pseudocode:
 
 ```python
+timestamp_reduced = block_header.timestamp % HISTORICAL_ROOTS_LENGTH
+timestamp_extended = timestamp_reduced + HISTORICAL_ROOTS_LENGTH
+timestamp_index = to_uint256_be(timestamp_reduced)
+root_index = to_uint256_be(timestamp_extended)
+
+timestamp_as_uint256 = to_uint256_be(block_header.timestamp)
 parent_beacon_block_root = block_header.parent_beacon_block_root
-timestamp = to_uint256_be(block_header.timestamp)
 
-sstore(HISTORY_STORAGE_ADDRESS, parent_beacon_block_root, timestamp)
+sstore(HISTORY_STORAGE_ADDRESS, timestamp_index, timestamp_as_uint256)
+sstore(HISTORY_STORAGE_ADDRESS, root_index, parent_beacon_block_root)
 ```
 
 #### New stateful precompile
 
-Beginning at the execution timestamp `FORK_TIMESTAMP`, the code and storage at `HISTORY_STORAGE_ADDRESS` constitute a "stateful" precompile.
+Beginning at the execution timestamp `FORK_TIMESTAMP`, a "stateful" precompile is deployed at `HISTORY_STORAGE_ADDRESS`.
+
+Callers of the precompile should provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
+Clients **MUST** sanitize this input call data to the precompile.
+If the input is _more_ than 32 bytes, the precompile only takes the first 32 bytes of the input buffer and ignores the rest.
+If the input is _less_ than 32 bytes, the precompile should revert.
```

Comment:
To clarify: if it reverts, would it just put 0 on the stack (signalling that the CALL has failed), instead of 1? Or, does it consume all provided gas? (In the earlier case the precompile always consumes 4200 gas)

Reply:
Regarding the reference implementation: seems like a 'normal' evm revert.

Reply:
the intention is a "normal" EVM revert

which I believe just puts 0 on the stack and consumes the 4200 gas

I don't see a reason for the precompile to consume all provided gas

If anything we could try to refine where execution fails and change how much gas is consumed but it is much simpler to just change the same amount no matter what

Reply:
Yup, I agree :+1: